### DISCUSSION TITLE
Usage of .chkp and field0.f0000* files.

### DISCUSSION THREAD

**Discussion by Bkv-1915** [ExtremeFLOW/neko#1626] [Q&A] [answered]
Hi creators,
I am new to Neko and I am working on turb_channel example. I want to know certain things:
1. Say I run simulation for 40 simulation time and get the .chkp file. How can i use this file as initial condition to run simulation from 40 - 80 simulation time and so on.
2. How to use these filed0.000* files or some other file (say .nek5000) to postprocess the results. Because, I am working on Ensemble Kalman filter and I need the values of the velocities. Where can I find these so that I can use this data on Kalman filter.
Thank you for your help

**Comment by timofeymukha** [comment]
Hi!

1. Yeah, that is what checkpoints are for. You can use the restart_file keyword in the case file and point it to your checkpoint. Take a look at the case file documentation, that should be there.
2. You can open the .nek5000 file in Paraview or Visit. There is also a Python package called pymech, which can open them.

**Comment by MartinKarp** [comment, marked as answer]
Information on restarts should be available here: https://neko.cfd/docs/develop/dd/d33/case-file.html

By setting the "restart_file" keyword you can restart your simulation.

As for the field files, you can load them into paraview or visit, or postprocess directly in fortran or us for example pymech to process the files in python https://pymech.readthedocs.io/en/stable/usage.html

Please observe that by default neko only writes the mesh information in the first field file, so the most reliable way to read the fiels is with the readnek function.

**Reply by Bkv-1915** [reply]
@MartinKarp , thank you for your kind reply. Regarding the fortran part, is there a usage manual for the same?

**Comment by Bkv-1915** [comment]
I want to read and modify the variables - u,v,w, p and T in the checkpoint files, so that I can restart my simulation with a slightly modified initial conditions. I was able to read the filed files using fortran and I was able to understand it a bit. However, I am not able to understand the structure of the checkpoint files. I am not able to extract the velocity components, pressure and temperature. Is there a manual for the checkpoint files?

**Reply by timofeymukha** [reply]
We have recently implemented the possibility to restart from a field file as well as the checkpoint. The checkpoints are designed to really preserve the entire state of the numerical solution, so you have e.g. lag arrays for velocity to directly start with a high order time integration. If you are going to mess around with values, all of that breaks anyway, so maybe restarting from a field file is a better alternative -- and those are easier to manipulate, as you already discovered.

**Reply by Bkv-1915** [reply]
Hi. I installed Neko-0.9.0, the latest version and tried to restart from the filed0.f00... files but failed. I have attached the errors. Please help me in resolving this. Thank you. 
![Neko_chkp_error](https://github.com/user-attachments/assets/33ea1ecf-0aed-4923-a136-73d79ba75dde)
![File_list](https://github.com/user-attachments/assets/2db36c32-0e23-459b-b3e5-f1e1e05d348b)
![Case_file](https://github.com/user-attachments/assets/bb853ef9-8ade-4481-8c97-e7afbc91b425)

**Reply by timofeymukha** [reply]
I led you astray a little bit. The restart_file keyword is only for checkpoints. But you can set the initial condition to be read from a field file. This just highlights the difference: an initial condition just sets the values of the unknowns, whereas a restart really fully loads the state of the simulation as saved in a checkpoint.

So, checkout Initial conditions in the case file documentation.
https://neko.cfd/docs/develop/dd/d33/case-file.html

**Reply by Bkv-1915** [reply]
@timofeymukha. Yes, its working now. Thank you.

**Reply by Bkv-1915** [reply]
Hi again. 

During the beginning of each simulation, the verbose that are printed on the terminal, ofcourse can be written on a .log file instead, takes 3-4 seconds (not sure if its the printing or its the entire initializing process). Can this be cut down, especially when I am restarting using a chkp and field file.  Because I need to Re-run the simulation for various inital conditions, say a 1000 times and its initializing each time which seems unnecessary. It would be great if you could redirect me to a source file where I can alter this if its not possible from the case file. I suppose, atleast the mesh file need not be Re-read. 

Im talking about this part (how much of it can be avoided if im restarting): 
![Screenshot from 2024-12-15 15-47-12](https://github.com/user-attachments/assets/6c4856d0-1fcc-4afb-b85b-fe995bfb2ec4)

**Reply by timofeymukha** [reply]
Not sure if I understand. The printing time is negligible, of course. I don't think there is anything you can skip. But perhaps I didn't understand what exactly you want to do.

**Reply by scarybenjz** [reply]
My apologies. 

In my project, I have to perform simulations for 1000 different initial conditions (specified by 1000 field files). Each simulation would run for the same simulation time, say from t_0 to t_end. I found out from the verbose that the actual time taken for each simulation in my case is 0.5 seconds, but the initialization (especially the gather-scatter part) takes 2.5 - 3 seconds. 

My question is this: is it possible to initialize the simulation (atleast the gather-scatter part) only for the first time instead of all 1000 simulations. Because this additional 3 seconds is costing me 40 days instead of 4 hours.

Somehow, if I could tell the neko code that the simulation is not terminated, but paused, and only the initial conditions are changed before it is resumed.

**Reply by timofeymukha** [reply]
OK. No, there is no way to do that through the case or user file. 

But why such a huge factor: 40 days vs 4 hours? The runtime is about 7 time longer, right? So, more like 30 hours vs 4?

**Reply by scarybenjz** [reply]
The simulation time, t_0 - t_end that I mentioned, its just one time-step. I have atleast 50 time-steps. For now. Down the line of my research work, I will have 10,000 different initial conditions, with 100 such time-steps.

**Reply by scarybenjz** [reply]
Initializing once would do wonders. It would cut-off this snowballing effect. 

It would be great if you could lead me to one or more of the files in the SRC folder. I think I can try adding a counter and a few if statements.

**Reply by timofeymukha** [reply]
I guess the more timesteps you have, the more negligible the init time, right? If one dt is 0.5s, then a hundred will be 50, dwarfing the 3s for the init?

The main time-loop is in simulation.f90 in neko_solve. I guess by then the init is done (please double-check). That would probably the way to introduce some hack. But consider an extra time whether you really want to go that way :-). Just so you don't spend more time coding than it would take to just run things as is.

**Reply by scarybenjz** [reply]
Is'nt it 300 sec for 100 dt. And I have 1000 simulation at each of this dt, so its 300,000 secs ?

**Reply by timofeymukha** [reply]
I guess my point is that the init with the GS you do only once in the beginning of eachof those 1000 sims. So the longer you run each sim, the more disregardable it becomes.

**Reply by pschlatt1** [reply]
do you want to run those simulations in serial (i.e. one after each other), or in parallel?

**Reply by scarybenjz** [reply]
@pschlatt1 I am running it on GPU and I have only one GPU. So in series.

**Reply by scarybenjz** [reply]
@timofeymukha But the GS is being executed at each of those 1000 sims. If its once, then my prob would be solved. Currently I am having a do loop to execute the mpirun command 1000 times. Now I feel that its the wrong way of doing it.

**Reply by pschlatt1** [reply]
Then my suggestion is to modify the code such that you run all your simuations in one "run", i.e. by replacing the initial conditions whenever a previous simulation is done. How exactly to do that in Neko I would not know (in Nek5000 it would be quite straightforward by just changing the temporal order to 1 and replacing the velocity field). Then you do not need to restart anything, mesh etc. is all the same.

**Reply by timofeymukha** [reply]
I still don't get it, to be honest. Either the init is a significant part of the total time, and then the total time has to be short, so the optimization is not worth it. Or each simulation is not short, and then the init is negligible... I guess I am missing something. Anyway, I guess you can try to hack your way in ;-).

**Reply by scarybenjz** [reply]
@timofeymukha  and @pschlatt1 , thank you for the quick replies. I appreciate it. I will now dive into Neko, hope I get this by this weekend. Fingers crossed.

**Reply by Bkv-1915** [reply]
Hi again. I was able to modify the code using your suggestions. Now the code is much faster as needed. I only made slight modification in the simulation.f90 file. i.e, I included another loop as shown in the screenshot to read field files and checkpoint files outside the time loop. 

What I want: When I do mpirun, it should read the case file once, and perform the simulation from t_0 to t_end several times ( modifying the initial condition according to the field files and reseting the time and time-step to zero)

I wanted to ask if additional modifications are required (screenshot included). 
![Screenshot from 2024-12-21 11-52-46](https://github.com/user-attachments/assets/8d59a2d3-8767-44ee-a2c7-6c85b6ceb17e)
